### PR TITLE
BL-2654 Fix IsConnectivityError

### PR DESCRIPTION
**Pull request by gmartin7**
IsConnectivityError was failing because of an && || precedence problem. Adding parentheses fixed it.

### CODE REVIEW

File: src/BloomTests/UpdateVersionTableTests.cs [hunk 1]
```diff
@@ -25,6 +25,16 @@ public void ThisVersionTooLarge_ReturnsEmptyString()
 		}
 
 		[Test]
+		public void UpgradeTableError_NoCrash()
+		{
+			var t = new UpdateVersionTable();
+			t.TextContentsOfTable = @"0.0.0,3.1.99999, http://first.com/first";
+			t.RunningVersion = Version.Parse("3.2.0");
+			var url = t.LookupURLOfUpdate();
```
Comment (JohnThomson) [inline src/BloomTests/UpdateVersionTableTests.cs]: presumably the steps above put the update lookup table in a state where Error is null, thus verifying that the second test of Error.Status does not unexpectedly happen.

Presumably, url (the result of LookupURLOfUpdate) is not actually a URL, but an UpdateTableLookupResult. And you're not testing an error state of UpgradeTable but its successful operation. But this is not obvious from the name of the variable or the name of the test method.

How about something like lookupResult as the variable name and for the method LookupURLOfUpdate_AllWell_ReportsNoConnectivityError?

Will this test fail if the computer running it is offline? Is it possible/worthwhile to do anything about it if so?

Comment (gmartin7) [inline src/BloomTests/UpdateVersionTableTests.cs]: I just ran it offline. Test passed.

File: src/BloomTests/UpdateVersionTableTests.cs [hunk 2]
```diff
@@ -25,6 +25,43 @@ public void ThisVersionTooLarge_ReturnsEmptyString()
...
+		}
+
+		[Test]
+		public void UpgradeTableParsingError_NoCrash()
+		{
+			var t = new UpdateVersionTable();
+			t.TextContentsOfTable = @"0.0.0,3,1,99999, http://first.com/first"; // too many commas
+			t.RunningVersion = Version.Parse("3.2.0");
+			var lookupResult = t.LookupURLOfUpdate();
+			Assert.IsNullOrEmpty(lookupResult.URL);
```
Comment (JohnThomson) [inline src/BloomTests/UpdateVersionTableTests.cs]: Do we also care that IsConnectivityError returns true?

File: src/BloomTests/UpdateVersionTableTests.cs [hunk 3]
```diff
@@ -25,6 +25,43 @@ public void ThisVersionTooLarge_ReturnsEmptyString()
...
+
+			t = new UpdateVersionTable();
+			t.TextContentsOfTable = @"0.0.0,, http://first.com/first"; // too few commas
+			t.RunningVersion = Version.Parse("3.2.0");
+			lookupResult = t.LookupURLOfUpdate();
+			Assert.IsNullOrEmpty(lookupResult.URL);
```
Comment (JohnThomson) [inline src/BloomTests/UpdateVersionTableTests.cs]: This looks like two independent tests (lookupURLOfUpdate_TooManyCommas_GivesErrorAndNoURL, etc)

File: src/BloomTests/UpdateVersionTableTests.cs [hunk 4]
```diff
@@ -25,6 +25,43 @@ public void ThisVersionTooLarge_ReturnsEmptyString()
...
+		}
+
+		[Test]
+		public void UpgradeTableVersionParsingError_NoCrash()
```
Comment (JohnThomson) [inline src/BloomTests/UpdateVersionTableTests.cs]: Could be renamed along lines above

**Comment by JohnThomson** [comment]
LGTM with minor suggestions, but I clearly don't fully understand the problem because I missed  a major problem before. Probably @hatton should take one more look.

File: src/BloomTests/UpdateVersionTableTests.cs (hunk 2 above)
Comment (gmartin7) [inline src/BloomTests/UpdateVersionTableTests.cs]: It doesn't.

File: src/BloomExe/UpdateVersionTable.cs [hunk 5]
```diff
@@ -81,21 +82,38 @@ public UpdateTableLookupResult LookupURLOfUpdate()
...
+					if (lower <= RunningVersion && upper >= RunningVersion)
+						return new UpdateVersionTable.UpdateTableLookupResult() { URL = parts[2].Trim() };
+				}
+			}
+			catch (ApplicationException e)
+			{
+				// BL-2654 Failure when reading upgrade table should not give a crash
+				// In this case, a line of the UpdateVersionTable was not parseable
+				// Put a message in the log and don't upgrade
+				Logger.WriteMinorEvent("Could not parse a line of the UpdateVersionTable" + e.Message);
```
Comment (hatton) [inline src/BloomExe/UpdateVersionTable.cs]: This is a major event. Minor events get lost unless an error happens really soon.

File: src/BloomExe/UpdateVersionTable.cs [hunk 6]
```diff
@@ -81,21 +82,38 @@ public UpdateTableLookupResult LookupURLOfUpdate()
...
+			}
+			catch (ArgumentException e)
+			{
+				// BL-2654 Failure when reading upgrade table should not give a crash
+				// In this case, a version number in the UpdateVersionTable was not parseable
+				// Put a message in the log and don't upgrade
+				Logger.WriteMinorEvent("Could not parse a version number in the UpdateVersionTable" + e.Message);
```
Comment (hatton) [inline src/BloomExe/UpdateVersionTable.cs]: Major event

**Comment by hatton** [comment]
BL-2654 called for toast. Where's the toast?

**Comment by gmartin7** [comment]
@hatton Does this generate a toast?
	Palaso.Reporting.ErrorReport.NotifyUserOfProblem(
		"Bloom failed to find if there is an update available, for some unknown reason.");
This scenario should trigger that (ApplicationUpdateSupport line 80)

File: src/BloomTests/UpdateVersionTableTests.cs (hunk 3 above)
Comment (gmartin7) [inline src/BloomTests/UpdateVersionTableTests.cs]: split apart

File: src/BloomTests/UpdateVersionTableTests.cs (hunk 4 above)
Comment (gmartin7) [inline src/BloomTests/UpdateVersionTableTests.cs]: done

**Pull request merged**